docs/service_architecture.md

```diff
@@ -0,0 +1,168 @@
+   for a signature, stores the newly signed timestamp in the database. It then
+   sends this new timestamp, along with the rest of the stored metadata, to the
+   requesting client.
+
+
+## Threat model
+
+Both the server and the signer are possible attack vectors. This section
+discusses how the architecture responds when compromised.
+
+## Notary server compromise
+
+In the event of a Notary server compromise, an attacker would have access to
+the metadata stored in the database as well as access to Notary signer to
+ sign anything with any key the Signer holds.
+
+- **Denial of Service** - An attacker could reject client requests and corrupt
+ 	or delete metadata from the database, thus preventing clients from being
+    able to download or upload metadata.
+
+- **Malicious Content** - An attacker can create, store, and serve arbitrary
+    metadata content for one or more repositories. However, they do not have
+    access to the original root, target, or (maybe) snapshots keys for
+    existing repositories.
+
+    Only clients who have never seen, and do not have any form of pinned trust
+    for, the compromised repositories can be tricked into downloading and
+    trusting the malicious content for these repositories.
+
+    Clients who have pinned trust for the compromised repositories, either
+    due to configuration or due to TOFU (trust on first use), will immediately
+    detect that the content is malicious and would not trust any root, targets,
+    or (maybe) snapshot metadata for these repositories.
+
+ - **Rollback, Mix and Match** - The attacker can request that
+    the Notary signer sign whatever timestamp (and maybe snapshot) metadata
+    they want. They can create valid timestamp and snapshot metadata that
+    certifies that the latest version of the repository contains
+    strictly older, or even a mix of older and newer, data.
+
+    Clients both with and without pinned trust would be vulnerable to these
+    attacks.
+
+ Note that the timestamp and snapshot keys are never compromised; once the
+ Server compromise is mitigated, an attacker will not be able to generate
+ valid timestamp or snapshot metadata and serve them on a malicious mirror, for
+ example.
+
+### Notary signer compromise
+
+In the event of a Notary signer compromise, an attacker would have access to
+all the private keys stored in a database. If the keys are stored in an HSM,
+they would have the ability to interfere with the keys in the HSM, but not
+to exfiltrate the private keys.
+
+- **Denial of Service** - An attacker could reject all Notary server requests
+  and corrupt or delete keys from the database (or even delete keys from an
+  HSM), and thus prevent Notary servers from being able to sign generated
+  timestamps or snapshots.
+
+- **Key Compromise** - If the Notary signer uses a database as its backend,
+  an attacker can exfiltrate all the private keys. This will let them set
```

Comment:
Technically they can also setup a malicious mirror to perform these attacks if they compromise the notary-server (since they can do arbitrary signatures).

I would remove the second sentence and add something to the effect of:

"Note that the capabilities of an attacker are the same as of a notary server compromise, with the important detail that in this particular case key rotation will be necessary to recover from the attack."